### PR TITLE
Migrate to Nunavut v2

### PR DESCRIPTION
**Pull request by Willmac16**
Addresses #277 

I'm still a bit unsure about how I'm handling `nunavut_support.py`, so I'd appreciate feedback on that and anything else.

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -59,7 +59,7 @@ packages             = find:
 # Think thrice before adding anything here, please.
 # The preferred long-term plan is to avoid adding any new required dependencies whatsoever for the project's lifetime.
 install_requires =
-    nunavut ~= 1.7
+    nunavut >= 2.0
```
Comment (pavel-kirienko) [inline setup.cfg:62]:
```suggestion
    nunavut ~= 2.3
```

File: tests/dsdl/_builtin_form.py
Comment (pavel-kirienko) [inline tests/dsdl/_builtin_form.py:1]: This test is no longer needed here as it has been migrated to Nunavut.

File: tests/dsdl/_constants.py
Comment (pavel-kirienko) [inline tests/dsdl/_constants.py:1]: This test is no longer needed here as it has been migrated to Nunavut.

File: tests/dsdl/_manual.py
Comment (pavel-kirienko) [inline tests/dsdl/_manual.py:1]: This test is no longer needed here as it has been migrated to Nunavut.

File: tests/dsdl/_rand.py
Comment (pavel-kirienko) [inline tests/dsdl/_rand.py:1]: This test is no longer needed here as it has been migrated to Nunavut.

File: tests/dsdl/_textual.py
Comment (pavel-kirienko) [inline tests/dsdl/_textual.py:1]: This test is no longer needed here as it has been migrated to Nunavut.

File: pycyphal/dsdl/_compiler.py [hunk 2]
```diff
@@ -81,9 +74,9 @@ def compile(  # pylint: disable=redefined-builtin
     A small subset of applications may require access to a generated entity without knowing in advance whether
     its name is a reserved identifier or not (i.e., whether it's stropped or not). To simplify usage,
     this submodule provides helper functions
```
Comment (pavel-kirienko) [inline pycyphal/dsdl/_compiler.py:76]:
```suggestion
    its name is a reserved identifier or not (i.e., whether it's stropped or not). To simplify usage,
    the Nunavut-generated module ``nunavut_support.py`` provides helper functions
```

**Review by pavel-kirienko — CHANGES_REQUESTED** [review]
Please run Black to fix code formatting.

Also do not forget to bump the minor version when the PR is ready.

**Comment by Willmac16** [comment]
It looks like Yakut needs the `pycyphal.dsdl.([a-z_]+)` -> `nunavut_support.$1` treatment.

**Comment by pavel-kirienko** [comment]
>Unrelated, should the package be marked as `zip_safe` in setup.cfg at this point?

Maybe, but it would probably be a micro-optimization and this flag is obsolete either way, so I prefer not to touch it. See https://setuptools.pypa.io/en/latest/deprecated/zip_safe.html

**Comment by pavel-kirienko** [comment]
>It looks like Yakut needs the pycyphal.dsdl.([a-z_]+) -> nunavut_support.$1 treatment.

I have opened a ticket about this in the Yakut repo. Would you be available to submit the required changes there? For testing, you can replace the PyCyphal dependency specification as follows:

```diff
 install_requires =
-    pycyphal[transport-udp,transport-serial,transport-can-pythoncan] ~= 1.8
+    pycyphal[transport-udp,transport-serial,transport-can-pythoncan] ~= git+https://github.com/Willmac16/pycyphal@Willmac16/issue277
```

You can temporarily replace the Yakut dependency in PyCyphal integration tests in a similar fashion.

File: pycyphal/application/_port_list_publisher.py [hunk 3]
```diff
@@ -10,13 +10,16 @@
 import pydsdl
 import pycyphal.application
 from pycyphal.transport import MessageDataSpecifier, ServiceDataSpecifier
+import pycyphal.dsdl
```
Comment (pavel-kirienko) [inline pycyphal/application/_port_list_publisher.py:13]: This import doesn't seem to be necessary.

File: tests/presentation/_pub_sub.py [hunk 4]
```diff
@@ -6,6 +6,7 @@
 import asyncio
 import pytest
 import pycyphal
+import nunavut_support
```
Comment (pavel-kirienko) [inline tests/presentation/_pub_sub.py:9]: Please move this import into the function scope.

File: tests/presentation/subscription_synchronizer/transfer_id.py [hunk 5]
```diff
@@ -11,6 +11,7 @@
 from pycyphal.transport.loopback import LoopbackTransport, LoopbackInputSession
 from pycyphal.presentation import Presentation
 from pycyphal.presentation.subscription_synchronizer.transfer_id import TransferIDSynchronizer
+import nunavut_support
```
Comment (pavel-kirienko) [inline tests/presentation/subscription_synchronizer/transfer_id.py:14]: Please move this import into the function scope.

**Review by pavel-kirienko — CHANGES_REQUESTED** [review]
>>It looks like Yakut needs the pycyphal.dsdl.([a-z_]+) -> nunavut_support.$1 treatment.
>
>I have opened a ticket about this in the Yakut repo.

Wait, I have a better idea. Can we please provide wrappers in `pycyphal.dsdl` (perhaps right in `__init__.py`, for simplicity) of the form:

```python
def update_from_builtin(destination: T, source: Any) -> T:
    """
    A wrapper over ``nunavut_support.update_from_builtin``.
    The ``nunavut_support`` module will be generated automatically if it is not importable.
    """
    from nunavut_support import update_from_builtin

    return update_from_builtin(destination, source)
```

That should avoid the breakage of Yakut.

-----

Also, could you please update the docs here:

- https://pycyphal--318.org.readthedocs.build/en/318/pages/architecture.html#dsdl-support
- https://pycyphal--318.org.readthedocs.build/en/318/api/pycyphal.dsdl.html#module-pycyphal.dsdl

`CompositeObject` and `ServiceObject` are simply gone now.

**Comment by Willmac16** [comment]
Once everything in this PR is good to go 🤞 

Can I squash my commits into one to keep the commit history of upstream clean.

File: docs/pages/architecture.rst [hunk 6]
```diff
@@ -243,13 +243,11 @@ The main API entries are:
...
   an instance of an autogenerated class.
 
-- :class:`pycyphal.dsdl.CompositeObject` and :class:`pycyphal.dsdl.ServiceObject` --- base classes for
-  Python classes generated from DSDL type definitions; message types and service types, respectively.
-
 - :func:`pycyphal.dsdl.to_builtin` and :func:`pycyphal.dsdl.update_from_builtin` --- used to convert
   a DSDL object instance to/from a simplified representation using only built-in types such as :class:`dict`,
   :class:`list`, :class:`int`, :class:`float`, :class:`str`, and so on. These can be used as an intermediate
   representation for conversion to/from JSON, YAML, and other commonly used serialization formats.
+  These functions are wrappers of the Nunavut generated support functions in `nunavut_support.py`.
```
Comment (pavel-kirienko) [inline docs/pages/architecture.rst:250]: This applies to `serialize`/`deserialize` as well.

File: pycyphal/dsdl/_compiler.py [hunk 7]
```diff
@@ -103,6 +96,7 @@ def compile(  # pylint: disable=redefined-builtin
         The source DSDL root namespace directory path. The last component of the path
         is the name of the root namespace. For example, to generate package for the root namespace ``uavcan``,
         the path would be like ``foo/bar/uavcan``.
+        If set to None, only nunavut_support will be generated
```
Comment (pavel-kirienko) [inline pycyphal/dsdl/_compiler.py:99]:
```suggestion
        If set to None, only the ``nunavut_support`` module will be generated.
```

File: pycyphal/dsdl/__init__.py [hunk 8]
```diff
@@ -17,29 +17,145 @@
...
-from ._composite_object import serialize as serialize
-from ._composite_object import deserialize as deserialize
+from ._import_hook import install_import_hook as install_import_hook
+import pydsdl
 
-from ._composite_object import get_fixed_port_id as get_fixed_port_id
-from ._composite_object import get_model as get_model
-from ._composite_object import get_class as get_class
-from ._composite_object import get_extent_bytes as get_extent_bytes
+T = TypeVar("T")
```
Comment (pavel-kirienko) [inline pycyphal/dsdl/__init__.py:28]: This should be either prefixed with an underscore or moved into a separate module to prevent scope leakage.

**Review by pavel-kirienko — CHANGES_REQUESTED** [review]
I think it looks good aside from a few minor nits. Please request another review when done and we'll have it merged.

**Review by pavel-kirienko — APPROVED** [review]

**Pull request merged**